pkg/build/controller/common/util.go

```diff
@@ -42,6 +48,85 @@ func HandleBuildCompletion(build *buildapi.Build, runPolicies []policy.RunPolicy
 	if err := runPolicy.OnComplete(build); err != nil {
 		utilruntime.HandleError(fmt.Errorf("failed to run policy on completed build %s/%s: %v", build.Namespace, build.Name, err))
 	}
+	if err := HandleBuildPruning(build, buildLister, buildConfigGetter, buildDeleter); err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to prune old builds %s/%s: %v", build.Namespace, build.Name, err))
+	}
+}
+
+func isCompletedBuild(build buildapi.Build) bool {
+	return build.Status.Phase == buildapi.BuildPhaseComplete
+}
+
+func isFailedBuild(build buildapi.Build) bool {
+	return build.Status.Phase == buildapi.BuildPhaseFailed
+}
+
+type ByStartTimestamp []buildapi.Build
+
+func (b ByStartTimestamp) Len() int {
+	return len(b)
+}
+
+func (b ByStartTimestamp) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+func (b ByStartTimestamp) Less(i, j int) bool {
+	return !b[j].Status.StartTimestamp.Time.After(b[i].Status.StartTimestamp.Time)
+}
+
+// HandleBuildPruning handles the deletion of old successful and failed builds
+// based on settings in the BuildConfig.
+func HandleBuildPruning(build *buildapi.Build, buildLister buildclient.BuildLister, buildConfigGetter buildclient.BuildConfigGetter, buildDeleter pruner.BuildDeleter) error {
+	bcName := buildutil.ConfigNameForBuild(build)
+
+	buildConfig, err := buildConfigGetter.Get(build.Namespace, bcName, metav1.GetOptions{})
```

Comment:
instead of going to the api to the get the buildconfig (and the builds), we should be going to the sharedinformer cache.  @csrwng might be able to give you some pointers.